### PR TITLE
Adding the ability to clear and show the pillar cache enabled when pillar_cache is True.

### LINKED ISSUE #37080: Clearing pillar cache from Master leaves stray files & does not appear to clear it
## Description of Issue/Question

Using the pillar cache with a disk backend. The `cache.clear_pillar` runner appears to clear the cache upon immediately calling `cache.pillar`, but the files still persist on the disk, and if it's within the cache TTL, the cached pillar data still gets retrieved, since those files still persist.

This is notable because we leverage a _custom external pillar_ that takes about 30 seconds to retrieve its data, so it's very obvious whether or not the cached data is being retrieved. I emphasized the fact that it is a custom external pillar in the event that this issue is a complication of using one.
### Setup

(Please provide relevant configs and/or SLS files (Be sure to remove sensitive info).)
Master config:

```
pillar_cache: True
pillar_cache_backend: disk
pillar_cache_ttl: 86400
```
### Steps to Reproduce Issue

(Include debug logs if possible and relevant.)
1) Populate the cache: `salt-call pillar.items`.
2) Attempts to clear cache with the runner:

``` bash
(salt) root@:pillar_cache # salt-run cache.clear_pillar '*'
True
(salt) root@:pillar_cache # salt-run cache.pillar
i-ot7rj6eg:
    ----------
i-v8vpgxba:
    ----------
(salt) root@:pillar_cache # ls -l
total 40
-rw-r--r-- 1 root root 17956 Oct 18 18:31 i-ot7rj6eg
-rw-r--r-- 1 root root 17956 Oct 18 18:54 i-v8vpgxba
```

3) Run `salt-call pillar.items` again.

```
[ERROR   ] Unable to read instance data, giving up
[INFO    ] Not an EC2 instance, skipping
[DEBUG   ] Connecting to master. Attempt 1 of 1
[DEBUG   ] Initializing new AsyncAuth for ('/etc/salt/pki/minion', 'i-v8vpgxba', 'tcp://172.254.254.100:4506')
[DEBUG   ] Generated random reconnect delay between '1000ms' and '11000ms' (9163)
[DEBUG   ] Setting zmq_reconnect_ivl to '9163ms'
[DEBUG   ] Setting zmq_reconnect_ivl_max to '11000ms'
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/minion', 'i-v8vpgxba', 'tcp://172.254.254.100:4506', 'clear')
[DEBUG   ] Decrypting the current master AES key
[DEBUG   ] Loaded minion key: /etc/salt/pki/minion/minion.pem
[DEBUG   ] SaltEvent PUB socket URI: /var/run/salt/minion/minion_event_4b91cc8893_pub.ipc
[DEBUG   ] SaltEvent PULL socket URI: /var/run/salt/minion/minion_event_4b91cc8893_pull.ipc
[DEBUG   ] Initializing new IPCClient for path: /var/run/salt/minion/minion_event_4b91cc8893_pull.ipc
[DEBUG   ] Sending event: tag = salt/auth/creds; data = {'_stamp': '2016-10-18T18:58:14.714370', 'creds': {'publish_port': 4505, 'aes': 'bWRJlUX08onQaBa1U4b9AV5JmEiS/QXwgZwWyNDrB1umL7ww38V2jr4NqNvDo1EUXQePf
PsAtDo=', 'master_uri': 'tcp://172.254.254.100:4506'}, 'key': ('/etc/salt/pki/minion', 'i-v8vpgxba', 'tcp://172.254.254.100:4506')}
[DEBUG   ] Loaded minion key: /etc/salt/pki/minion/minion.pem
[DEBUG   ] Determining pillar cache
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/minion', 'i-v8vpgxba', 'tcp://172.254.254.100:4506', 'aes')
[DEBUG   ] Initializing new AsyncAuth for ('/etc/salt/pki/minion', 'i-v8vpgxba', 'tcp://172.254.254.100:4506')
[DEBUG   ] Loaded minion key: /etc/salt/pki/minion/minion.pem
[DEBUG   ] LazyLoaded jinja.render
[DEBUG   ] LazyLoaded yaml.render
[DEBUG   ] LazyLoaded pillar.items
[DEBUG   ] Determining pillar cache
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/minion', 'i-v8vpgxba', 'tcp://172.254.254.100:4506', 'aes')
[DEBUG   ] Initializing new AsyncAuth for ('/etc/salt/pki/minion', 'i-v8vpgxba', 'tcp://172.254.254.100:4506')
[DEBUG   ] Loaded minion key: /etc/salt/pki/minion/minion.pem
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/minion', 'i-v8vpgxba', 'tcp://172.254.254.100:4506', 'aes')
[DEBUG   ] Initializing new AsyncAuth for ('/etc/salt/pki/minion', 'i-v8vpgxba', 'tcp://172.254.254.100:4506')
[DEBUG   ] LazyLoaded nested.output
```

4) Repeat step 2, then delete the files in `pillar_cache`.
5) The pillar refreshes its data from the external pillar as expected. The `SaltReqTimeoutErrors` are safe to ignore.

```
[ERROR   ] Unable to read instance data, giving up
[INFO    ] Not an EC2 instance, skipping
[DEBUG   ] Connecting to master. Attempt 1 of 1
[DEBUG   ] Initializing new AsyncAuth for ('/etc/salt/pki/minion', 'i-v8vpgxba', 'tcp://172.254.254.100:4506')
[DEBUG   ] Generated random reconnect delay between '1000ms' and '11000ms' (10703)
[DEBUG   ] Setting zmq_reconnect_ivl to '10703ms'
[DEBUG   ] Setting zmq_reconnect_ivl_max to '11000ms'
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/minion', 'i-v8vpgxba', 'tcp://172.254.254.100:4506', 'clear')
[DEBUG   ] SaltReqTimeoutError, retrying. (1/7)
[DEBUG   ] SaltReqTimeoutError, retrying. (2/7)
[DEBUG   ] SaltReqTimeoutError, retrying. (3/7)
[DEBUG   ] SaltReqTimeoutError, retrying. (4/7)
[DEBUG   ] SaltReqTimeoutError, retrying. (5/7)
[DEBUG   ] SaltReqTimeoutError, retrying. (6/7)
[DEBUG   ] SaltReqTimeoutError, retrying. (7/7)
```
### Versions Report

(Provided by running `salt --versions-report`. Please also mention any differences in master/minion versions.)

Masters and minions use the same version.

```
Salt Version:
           Salt: 2016.3.3-186-g86ac8bd

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.5.3
          gitdb: 0.6.4
      gitpython: 2.0.2
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.9 (default, Jun 29 2016, 13:08:31)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 16.0.0
           RAET: Not Installed
          smmap: 0.9.0
        timelib: Not Installed
        Tornado: 4.4.2
            ZMQ: 4.1.5

System Versions:
           dist: debian 8.6 
        machine: x86_64
        release: 3.16.0-4-amd64
         system: Linux
        version: debian 8.6
```

### PR DESCRIPTION
**Pull request by garethgreenaway**
### What does this PR do?
Adding the ability to clear and show the pillar cache enabled when pillar_cache is True.

### What issues does this PR fix or reference?
Fixes:  #37080 

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [ ] Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

File: salt/runners/pillar.py [hunk 1]
```diff
@@ -97,3 +98,104 @@ def show_pillar(minion="*", **kwargs):
...
+
+        if tgt in _pillar_cache and _pillar_cache[tgt]:
+            pillar_cache[tgt] = _pillar_cache.get(tgt).get(pillarenv)
+
+    return pillar_cache
+
+
+def show_pillar_cache(minion="*", **kwargs):
+    """
+    Clears the cached values when using pillar_cache
```
Comment (Ch3LL) [inline salt/runners/pillar.py:156]: This looks like the same description from the clear_pillar_cache function

File: salt/runners/pillar.py [hunk 2]
```diff
@@ -97,3 +98,104 @@ def show_pillar(minion="*", **kwargs):
...
+
+    ckminions = salt.utils.minions.CkMinions(__opts__)
+    ret = ckminions.check_minions(minion)
+
+    pillarenv = kwargs.pop("pillarenv", None)
+    saltenv = kwargs.pop("saltenv", "base")
+
+    pillar_cache = {}
+    for tgt in ret.get("minions", []):
+        saltenv = "base"
```
Comment (Ch3LL) [inline salt/runners/pillar.py:180]: Why are you setting saltenv to base right after getting it from kwargs. Same with the function above.

File: tests/unit/runners/test_pillar.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
+"""
```
Comment (Ch3LL) [inline tests/unit/runners/test_pillar.py:1]: can you convert to pytest test.

**Review by Ch3LL — CHANGES_REQUESTED** [review]

File: salt/runners/pillar.py (hunk 1 above)
Comment (twangboy) [reply salt/runners/pillar.py:156]: Yep, should be `Shows the cached values...`

**Review by twangboy — CHANGES_REQUESTED** [review]

**Review by twangboy — DISMISSED** [review]

File: salt/runners/pillar.py [hunk 4]
```diff
@@ -97,3 +98,103 @@ def show_pillar(minion="*", **kwargs):
...
+            __opts__, grains, id_, saltenv, pillarenv=pillarenv
+        )
+
+        if __opts__.get("pillar_cache_backend") == "memory":
+            _pillar_cache = pillar.cache
+        else:
+            _pillar_cache = pillar.cache._dict
+
+        if tgt in _pillar_cache and _pillar_cache[tgt]:
+            pillar_cache[tgt] = _pillar_cache.get(tgt).get(pillarenv)
```
Comment (s0undt3ch) [inline salt/runners/pillar.py:198]:
```suggestion
            pillar_cache[tgt] = _pillar_cache[tgt].get(pillarenv)
```
At this stage we now `tgt` is in the pillar cache.

**Review by s0undt3ch — DISMISSED** [review]
Left a change but it's not a blocker

**Review by s0undt3ch — DISMISSED** [review]

File: salt/runners/pillar.py [hunk 5]
```diff
@@ -97,3 +98,103 @@ def show_pillar(minion="*", **kwargs):
...
+
+    ckminions = salt.utils.minions.CkMinions(__opts__)
+    ret = ckminions.check_minions(minion)
+
+    pillarenv = kwargs.pop("pillarenv", None)
+    saltenv = kwargs.pop("saltenv", None)
+
+    pillar_cache = {}
+    for tgt in ret.get("minions", []):
+        saltenv = "base"
```
Comment (Ch3LL) [inline salt/runners/pillar.py:129]: pretty sure this saltenv needs to be cleaned up as well, since we are grabbing from kwargs above.

**Review by Ch3LL — CHANGES_REQUESTED** [review]

File: salt/runners/pillar.py (hunk 5 above)
Comment (garethgreenaway) [reply salt/runners/pillar.py]: Fixed.

File: salt/runners/pillar.py [hunk 6]
```diff
@@ -97,3 +98,102 @@ def show_pillar(minion="*", **kwargs):
 
     compiled_pillar = pillar.compile_pillar()
     return compiled_pillar
+
+
+def clear_pillar_cache(minion="*", **kwargs):
+    """
+    Clears the cached values when using pillar_cache
+
```
Comment (s0undt3ch) [inline salt/runners/pillar.py:106]: `.. versionadded:: 3003`?

File: salt/runners/pillar.py [hunk 7]
```diff
@@ -97,3 +98,102 @@ def show_pillar(minion="*", **kwargs):
...
+        if tgt in _pillar_cache and _pillar_cache[tgt]:
+            pillar_cache[tgt] = _pillar_cache.get(tgt).get(pillarenv)
+
+    return pillar_cache
+
+
+def show_pillar_cache(minion="*", **kwargs):
+    """
+    Shows the cached values in pillar_cache
```
Comment (s0undt3ch) [inline salt/runners/pillar.py:155]: Also version added...

File: tests/pytests/unit/runners/test_pillar.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+            MagicMock(return_value=_CHECK_MINIONS_RETURN),
+        ), patch("salt.utils.atomicfile.atomic_open", mock_open()) as atomic_open_mock:
+            with patch("os.path.exists", MagicMock(return_value=True)):
+                with patch(
+                    "salt.utils.files.fopen", mock_open(read_data=cache_contents)
+                ) as fopen_mock:
+                    # No arguments defaults to all globbing
+                    ret = pillar_runner.clear_pillar_cache()
+                    assert ret == {}
+
```
Comment (s0undt3ch) [inline tests/pytests/unit/runners/test_pillar.py:75]: Isn't the cache already clear at this point?

**Review by s0undt3ch — DISMISSED** [review]

File: salt/runners/pillar.py (hunk 6 above)
Comment (garethgreenaway) [reply salt/runners/pillar.py:106]: Done

File: salt/runners/pillar.py (hunk 7 above)
Comment (garethgreenaway) [reply salt/runners/pillar.py:157]: Done.

File: tests/pytests/unit/runners/test_pillar.py (hunk 8 above)
Comment (garethgreenaway) [reply tests/pytests/unit/runners/test_pillar.py]: At that point the cache has data but it's not obvious from the tests, so I rewrote them.

File: salt/runners/pillar.py [hunk 9]
```diff
@@ -97,3 +98,106 @@ def show_pillar(minion="*", **kwargs):
 
     compiled_pillar = pillar.compile_pillar()
     return compiled_pillar
+
+
+def clear_pillar_cache(minion="*", **kwargs):
+    """
+    Clears the cached values when using pillar_cache
+
+    .. versionadded:: 3003.3
```
Comment (s0undt3ch) [inline salt/runners/pillar.py:107]:
```suggestion
    .. versionadded:: 3003
```

File: salt/runners/pillar.py [hunk 10]
```diff
@@ -97,3 +98,106 @@ def show_pillar(minion="*", **kwargs):
...
+            pillar_cache[tgt] = _pillar_cache.get(tgt).get(pillarenv)
+
+    return pillar_cache
+
+
+def show_pillar_cache(minion="*", **kwargs):
+    """
+    Shows the cached values in pillar_cache
+
+    .. versionadded:: 3003.3
```
Comment (s0undt3ch) [inline salt/runners/pillar.py:159]:
```suggestion
    .. versionadded:: 3003
```

**Review by s0undt3ch — CHANGES_REQUESTED** [review]
😄

File: salt/runners/pillar.py (hunk 9 above)
Comment (garethgreenaway) [reply salt/runners/pillar.py]: Bah.  Fixed. 😂

**Review by s0undt3ch — APPROVED** [review]

**Review by Ch3LL — APPROVED** [review]

**Pull request merged**